add EP for environment variable API (#726)
shipwright-io/build: docs/proposals/env-vars.md

	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
}
```

This proposal is making the simplifying assumption that ultimately setting the `Env []EnvVar` field in the 
underlying Tekton `Step`/`Container` is sufficient for an initial implementation and scenarios we want to support.
That will help the initial implementation in that the explicit "name" index field helps with default kubernetes 
"path" API operations that tend to occur with devops/gitops flows.

Next, we'll couple the default value capability available in Tekton's `StepTemplate` with reuse of the k8s
`EnvVar` type.  In other words, a default value field can be mapped to a `StepTemplate`'s env var setting.  

With that, a new Shipwright type would be:

```go
type EnvironmentVariable struct {
   corev1.EnvVar
   
   // DefaultValue if set is the value applied if neither EnvVar.Value or EnvVar.ValueFrom have a value even 
   // after any Parameter variable substitution.
   // +optional
   DefaultValue  string
}
```

Then, the specs of all `BuildStrategy` (both cluster and namespaced scope), `Build`, and `BuildRun` will be 
updated with an array of `EnvironmentVariable`

> `BuildStrategy` would also have a `EnvironmentVariable` array named `DisallowedEnvironmentVariables` that lists `EnvVar` settings which are not allowed

Reviewer: Thinking out loud, I wonder if we should invert this logic and let administrators define "allowed environment variables" instead.  Like for instance `HTTP_PROXY`, an administrator would like to leave a list of proxy servers that are allowed; in comparison, would be difficult to add a list of servers that aren't allowed.

Author: Yeah I could see both allow and deny lists being desirable @otaviof .... I tried to indicate that providing some form of both may be a item to strive fo, though maybe we stage one before the other wrt to the notion of a staging plan I proposed in response to some of @SaschaSchwarze0 's comments.  Though based on some of the exchanges up above, I have some optimism we devise and order of precedence that gets us suitable allow/deny capabilities.

Bottom line though, I need to add more details to the portion you've highlighted here.

and will be filtered in the final application onto the `TaskSpec`.

With settings at each of those "layers", order of precedence comes in.

A proposed default would be "developer centric":
- apply `BuildStrategy` envs first
- then `Build`
- then `BuildRun`

Then, via the [current global config of controller env vars](https://github.com/shipwright-io/build/blob/v0.4.0/docs/configuration.md),

> or if that current approach gets converted to a `ConfigMap`, or if the community decides to create a `ConfigMap` for each

Reviewer: Would you please add moro on why a `ConfigMap` would be useful/needed on this process?

Author: I'd rather not detail why that shift might occur as part of this EP, but defer to whichever EP is created to dive into that, if the change in fact occurs.

https://github.com/shipwright-io/build/issues/651 was opened an issue to track this potential change.  I'll add a qualifier here to reference that issue and potential/subsequent EP for details.

That sound OK to you @otaviof ?

Reviewer: Yes, it does. I'll watch out for #651, thanks!

example.

That said, one could also see cluster administrators wanting certain environment variables to never be set in their
cluster, don't use certain values, or that certain environment variables always use certain values.  
Perhaps certain values for an environment variable, or the existence of the environment variable at all,
enable features in underlying tools encapsulated by a given `BuildStrategy` that are deemed inappropriate, unsafe,
of too costly resource wise for the cluster in question.

There are a few possibilities in how the API is or is not structured that could allow for support of each of the above
concerns.  We should get to a consensus on which way(s) we support in the enhancement proposal. Ideally, one approach
makes for a simpler implementation path.  If multiple approaches seem necessary, then we have to sort out if those 
varying approaches can be staged, or have to be implemented all at once.

## Summary

One of the fields in the core Kubernetes API, Pods and their Containers, are Environment Variables.  Of course those
translate to operating system level environment variables being set when running the same commands executed

> in Pod Containers from the command line of their personal computuer.

Reviewer 2: ```suggestion
in Pod Containers from the command line of their personal computer.
```

The next major dependency for Shipwright, Tekton, has a vast enough array of scenarios it addresses that it provides a 
direct line to the Kubernetes Container API in their API.  As as result, environment variable fields from `Steps`/`Containers`
are directly accessible from Tekton API.

> Specifically [Tekton TaskRuns](https://github.com/tektoncd/pipeline/blob/main/docs/taskruns.md) reference

Reviewer 2: Steps are defined in the spec of a Task, not a TaskRun.

Author: yes but there is a ptr to a TaskSpec  in the TaskRun, and the Steps are in the TaskSpec definition.  

if that is not clear my text above I'll add more verbiage

Author: ok finally back to this today .... even if you and I are okay,  I'm making the relationship chain more explicit and clear in my next update

used in Shipwright.

So supporting our users via our API and related features to manage environment variables for those tools makes a lot of
sense.

### Goals

- Shipwright API needs to articulate a first class way for specifying those environment variables approach for both 
  administrators and individual developers (i.e. authors/owners of the BuildStrategies vs. authors owners of the Builds 
  and BuildRuns) to specify those environment variables.
- Use of the existing features Tekton provides to enrich our users' management of environment variables needs to be 
  evaluated and exploited for code reuse.
- Maintain the currently Shipwright approach of treating Tekton as an implementation detail, and try to not leak Tekton
  API into our API.

### Non-Goals

- Any needed change to integrate Shipwright with Tekton parameters is handled by the [current 'Parameterize Build Strategies' EP](https://github.com/shipwright-io/build/pull/697)
  or follow up work that stems from that EP.  Until then, only static content of any environment variable value related 
  fields is supported.  In other words, the `$(params...)` syntax seen in Tekton today.

> 

Reviewer 2: To make it clear, it may make sense to add a statement that this EP is not about environment variables that will be set in the final container image.

Author: sure that distinction has value ... I'll add

	// Source for the environment variable's value. Cannot be used if value is not empty.
	// +optional
	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
}
```

This proposal is making the simplifying assumption that ultimately setting the `Env []EnvVar` field in the 
underlying Tekton `Step`/`Container` is sufficient for an initial implementation and scenarios we want to support.
That will help the initial implementation in that the explicit "name" index field helps with default kubernetes 
"path" API operations that tend to occur with devops/gitops flows.

Next, we'll couple the default value capability available in Tekton's `StepTemplate` with reuse of the k8s
`EnvVar` type.  In other words, a default value field can be mapped to a `StepTemplate`'s env var setting.  

With that, a new Shipwright type would be:

```go
type EnvironmentVariable struct {
   corev1.EnvVar
   
   // DefaultValue if set is the value applied if neither EnvVar.Value or EnvVar.ValueFrom have a value even 
   // after any Parameter variable substitution.
   // +optional
   DefaultValue  string
}
```

> Then, the specs of all `BuildStrategy` (both cluster and namespaced scope), `Build`, and `BuildRun` will be 

Reviewer 2: In the BuildRun, there would not anymore be a need for a default value. At the latest at this point, the value needs to be specified by the user.

Author: reusing the type across the board is attractive to me here

a bit of a stretch, but if a user was uncertain about the contents of a configmap or secret used with valueFrom (perhaps they reconcile as blank) having the default could a safe fallback

so I'm inclined to keep as is ... but if that reasoning is not sufficient for you, please elaborate and we'll go from there

Author: So a lot of good discourse to unpack there @SaschaSchwarze0.  Bottom line, I believe I need to make some adjustments.  

But how to convey those so we can reach consensus?  Here we go ... but while I was finishing this post, @SaschaSchwarze0 and I chatted on slack and also agreed to set up a video conf for us to discuss.
So while we can all continue to talk here (I also have comments from @otaviof I need to get to), this can also serve as additional input into the conversation he and I will have.

1) Overall, the notion of conveying a staging plan wrt this EP is a possibility.  I think we could incrementally add to an env var API as part of Shipwright if need be.  I'll circle back to that based on where we land.  Then, if need be, a stage 1 could be simple k/v pairing, a stage 2 maybe could be say `valueFrom` (though I think it should be a stage 1 item, with possible pull out if something unexpected arises), followed by a probable `no` for `defaults`.  More on defaults in a bit, but perhaps I've overstretched wrt that.

2) IMO, we do have uses cases for `valueFrom`, again with the caveat that we do not have to necessarily address them in an "initial" stage if when we get to implementation there is uncertainty on the nature of the change.

Those use cases?

First, I already have a link for this in this EP, but Tekton's specific example at https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#using-a-secret-as-an-environment-source where they seed the `GITHUB_TOKEN` based on what is pulled from a k8s secret, is pervasive enough IMO that I at least could see Shipwright users liking this form.

A lot of tools like git, maven, etc. used in building are big on environment variables, and in the case of sensitive data, storing that data and retrieving it from k8s Secrets is a de facto mode of operator at this point.

And a first blush,  it seems like the reconciliation of the `BuildRun` to `TaskRun` is straight forward, as we map the analogous `BuildRun` API to the `TaskRun` API seen in that link.

Hopefully with that, along with cleaning things up around the strategy already has steps/containers point you made (more on that below), that will remove the complexity concern.

But I think it is fair to say that a path to Shipwright will be Tekton users that desire the value in standardizing image building.  And even though Tekton is a Shipwright implementation detail, analogous semantics like this, that leverage core k8s resources, seems like a distinct possibility.

Second, we also have this analogous scenario wrt build tools and this precise feature with OpenShift Builds.  Please see https://docs.openshift.com/container-platform/4.7/cicd/builds/creating-build-inputs.html#builds-using-secrets-as-environment-variables_creating-build-inputs

And from a historical perspective, that feature arose from customer feedback (again, how they are use to supplying tokens to build tools via env vars).  Long term, as we migrate OpenShift Build users to Shipwright, not having this could be a feature gap and obstruction to migration.  

Now, back to the staging, again, perhaps we do not do this initially based on how the initial implementation goes, but I think we should strive for doing it initially, and only back it out if we see encounter something unexpected in mapping from BuildRun to TaskRun during the implementation PR.

But minimally though, if we don't have direct mapping of this features for these two potential user sets, we then have a more complex migration of "how you do this in a different form via Shipwright" story that we'll need to convey.

3) Defaulting ... yeah, my motivation here stemmed from my exploration of Tekton Triggers, and their use of defaults with the parameters for `TriggerTemplates`.  Fixed some bugs in this space, and overall the feature resonated with me.

That said, none of the existing stakes in the ground wrt environment variables I noted above applied the defaulting concept.

Based on that and your initial interpretation, I'm going to back off of it in the main proposal section.  I'll probably add some mention of it in alternatives, where I'll mention the idea was entertained, but we moved off of it.

4) How it comes together .... strategy already having steps/containers.  Yeah I missed the need for more detail here, thanks.  And I think your comment hit on the choice we have here.

Setting env var's on the build step(s) is already available, so is that sufficient for setting things at the `BuildStrategy` level?  
Do we allow setting of env vars at the `BuildStrategySpec` level, that apply to all steps/containers?

I could see staging coming into play here, where we at least initially we do not add a new field for multi container override, and then wait for user feedback.

Then, separate from those choices, the precedence decision I've noted in the doc, where the admin and dev sort out whether the `BuildStrategy` or `BuildRun` take precedence, or if we do not provide that choice and decide on an order, at least initially builds on top of those choices above.

Controlling the precedence order could also be another staging item.  Where initially we just pick an order, perhaps "developer" focused, where the BuildRun wins, and then we let administrators block that with kyverno / gatekeeperm though I could go the other way as well, and then add configuration / control around the order later.

updated with an array of `EnvironmentVariable`

`BuildStrategy` would also have a `EnvironmentVariable` array named `DisallowedEnvironmentVariables` that lists `EnvVar` settings which are not allowed

> and will be filtered in the final application onto the `TaskSpec`.

Reviewer 2: Without having deeply thought about this ...

* black listing might also be something on the deployment level.
* white listing might be interesting as well, for example in Paketo, environment variables are typically prefixed with `BP`

You may also add some words about what happens if a Build or BuildRun contains some environment variables that are not allowed.

Author: ah yes I only implied here that filtering would occur from the unallowed list

do we update status/conditions?
do we add logs / events?
do we prevent execution (seems extreme ... but maybe an opt in for the administrator)?

I'll ponder and update, thanks.  Of course chime in with any immediate opinions on how to answer those questions.

Reviewer 2: I think the behavior should be the same as for params, there I had this exchange with Enrique: https://github.com/shipwright-io/build/pull/697/files#r604954782. So, I in general prefer that we do not just silently filter. I consider writing to the controller's container log or events silently because that's not where a build user would necessarily look at. I would prefer the validation to happen as early as possible - with webhooks, this would means to prevent the creation of the build or buildrun. With our current approach, their status would be not registered or not succeeded with reasonable reasons. Anyway, when we create the TaskRun, we always need to validate again because the build strategy might have changed after a build was created (and successfully validated by an eventually existing webhook or registered by today's approach).

Author: I need to re-review the merged parameters EP in details, but on the surface, I'm good with being consistent and updating this EP as needeed for consistency.  

+1 on non-silent filtering, the earlier the better, all that.

@SaschaSchwarze0 and I can dive into this more when we chat.

With settings at each of those "layers", order of precedence comes in.

A proposed default would be "developer centric":
- apply `BuildStrategy` envs first
- then `Build`
- then `BuildRun`

Then, via the [current global config of controller env vars](https://github.com/shipwright-io/build/blob/v0.4.0/docs/configuration.md),
or if that current approach gets converted to a `ConfigMap`, or if the community decides to create a `ConfigMap` for each
logical piece of "feature configuration", we have a setting that allows a switch to "ADMINISTRATOR_CENTRIC" order of precedence,
say:

- apply `BuildRun` envs first
- then `Build`

> - then `BuildStrategy`

Reviewer 2: What would be the use case for this?

Author: The use case I mentioned above that the administrator wants to enforce specific settings for env vars

Reviewer 2: Good, agree. For me this is not a global configuration but rather a case-by-case decision for every environment variable. Earlier I commented that we will need to distinguish - in the build strategy - between `spec.env` (what you propose to add) vs `spec.buildSteps[].env`. A reasonable approach imo would be that those environment variables explicitly defined on a step cannot be overwritten. An example would be the `BUILDKITD_FLAGS` environment variable of the BuildKit build strategy.

Author: Yeah perhaps avoiding a global config and supporting case by case would be an argument for having both `spec.env` and `spec.buildSteps[].env` .... where the latter is the final override, and the former is the "set up an initial value, but allow the `BuildRun` to override.

We can settle on this during our discussion @SaschaSchwarze0

> 

Reviewer 2: Looking for some more information on where we set those environment variables eventually. I assume on all steps defined in the build strategy = not including those steps/containers that we dynamically define like to retrieve sources ?

Author: Yeah I can add more details here based on how we resolve items above after our conversation.

- `BuildStep` from the `BuildStrategySpec` is next.  The key/value pair from these override anything from
  `BuildStrategySpec.[]EnvironmentVariable`, `BuildSpec.[]EnvironmentVariable`, `BuildRunSpec.[]EnvironmentVariable`
- `BuildRunSpec.[]EnvironmentVariable` is next
- followed by `BuildSpec.[]EnvironmentVariable`
- and lastly, `BuildStrategySpec.[]EnvironmentVariable`

Or conversely, the environment variables are applied in this fashion
- first from `BuildStrategySpec.[]EnvironmentVariable` assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used
- then from `BuildSpec.[]EnvironmentVariable`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
  where values are replaced from any entries previously seeded
- then from `BuildRunSpec.[]EnvironmentVariable`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
  where values are replaced from any entries previously seeded
- then from `BuildStrategySpec.[]BuildStep`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,

>   where values are replaced from any entries previously seeded  

Reviewer 3: I was thinking of another approach:

* Everything specified in `BuildStrategySpec.[]BuildStep` may be overriden in Build/BuildRun unless specified otherwise in `.spec.[]DisallowEnvironmentVariableOverride`.

*  `.spec.[]DisallowEnvironmentVariableOverride` would contain the list of environment variables that a user is not allowed to be set irrespective of whether there was a default specified in `BuildStrategySpec.[]BuildStep`.
* `.spec.[]EnvironmentVariable` is not needed anymore because defaults could be specified in the `BuildStep` itself. 
 
( did you intend `BuildRunSpec.[]EnvironmentVariable` to be an allow-list of what env vars can be specified in build/buildRun ?)

Author: yep good question @sbose78 .... need some more  work here
- `BuildRunSpec.[]EnvironmentVariable` in intended to be defaults essentially
- but that said, admins might want to start with an "allow no envs, then list a small set of envs that can be used" vs. "allow all evns, then lis a small set of envs that cannot be used"

I'll add a `.spec.[]AllowedEnvironmentVariables` field, along with some thought / explanation on how `Allow...` and `Disallow...` work together.

Spoiler: `Allow...` and "allow no envs, then list a small set of envs that can be used" will take precedence.

Reviewer 3: Agreed @gabemontero , sounds good!

Author: update pushed @sbose78

```yaml
apiVersion: shipwright.io/v1alpha1
kind: ClusterBuildStrategy
metadata:
  name: something-that-uses-docker
spec:
  # nobody is allowed to change DOCKER_TLS_VERIFY
  disallowedEnvironmentVariableKeys:
    - 'DOCKER_TLS_VERIFY'
  environmentVariables:
    # for convenience, we add the version most our users can use, but we'll allow you to override
    - name: DOCKER_API_VERSION
      value: '1.19'
  buildSteps:
    - command:
        - /usr/local/bin/docker
        - build
      env:
        - name: DOCKER_CONFIG
          value: /tekton/home/.docker

```

`Build` with an override of `DOCKER_API_VERSION`:

```yaml
apiVersion: shipwright.io/v1alpha1
kind: Build
metadata:
  name: something-that-uses-docker-build
spec:
  source:
    url: https://github.com/myorg/myrepo
  strategy:
    name: something-that-uses-docker
    kind: ClusterBuildStrategy
  environmentVariables:
    - name: DOCKER_API_VERION
      value: '1.20'
```

`BuildRun` with an override of `DOCKER_API_VERSION`:

```yaml
apiVersion: shipwright.io/v1alpha1
kind: BuildRun
metadata:
  name: something-that-uses-docker-buildrun
spec:
  environmentVariables:

>     - name: DOCKER_API_VERSION

Reviewer 3: Is this the only environment variable the Build/BuildRun is allowed to set in this example ?

Author: nope per our exchange below

<!--
Copyright The Shipwright Contributors

SPDX-License-Identifier: Apache-2.0
-->

---
title: Expressing Environment Variables in Shipwright API
authors:
  - "@gabemontero"

reviewers:
  - "@sbose78"
  - "@SaschaSchwarze0"
  - "@otaviof"
<!-- 
Note from Gabe: I went light (i.e. just Shoubhik since he asked for this EP) on reviewers as a first pass,
only because it implied a mandate / expectation.  I've chosen to update this list and folks actually review.
-->

approvers:
  - "@sbose78"
<!-- 
Note from Gabe: I went light (i.e. just Shoubhik since he asked for this EP) on approvers as a first pass,
only because it implied a mandate / expectation.  I've chosen to update this list and folks with approve permissions actually review,
where I'll certainly make sure we get multi company sign off.
-->

creation-date: 2021-04-09

last-updated: 2021-04-27

status: implementable    

<!-- status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced -->
see-also:
  - [Parameterize Build Strategies](https://github.com/shipwright-io/build/pull/697/files)
---

# Expressing Environment Variables in Shipwright API

## Release Signoff Checklist

> - [ ] Enhancement is `implementable`

Author: updating the check list will be part of the next push

- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [docs](/docs/)

## Open Questions [optional]

> 1. API change

Assuming this gets delivered while Shipwright builds is still v1alpha1, we just have to announce to users
what new API is provided.

> But if the implementation of this is delayed to the point that we are at say v1beta1, then we have to 

Reviewer 4: I don't think we can confidently ship a v1beta1 API without this implemented.

Author: part of next push

consider the round trip experience when a client is at a different version of the server as discussed in
[Kubernetes Deprecation Policy Rule #2](https://kubernetes.io/docs/reference/using-api/deprecation-policy/).
And yes, adding an API is not the same as deprecating one, but the round trip guarantee applies (as Tekton
learnt recently).  The use of an annotation that captures version differences comes into play.

> 2. Order of precedence

Similar to other shipwright enhancements currently in flight, the notion of settings at a `BuildStrategy` level
serving as a default, with the ability to override those defaults at a more granular `Build` or `BuildRun`, resonates
with our topic here of managing environment various in Shipwright.  This order makes sense with developer flows for
example.

That said, one could also see cluster administrators wanting certain environment variables to never be set in their

> cluster, don't use certain values, or that certain environment variables always use certain values.  

Author: My initial thought here was the black box nature of openshift builds actually shielded its exposure of environment variables from some of the more complex concerns that can occur as the box is opened up with tekton based shipwright.

I agree with cycling through each feature / use case on its own to see if it remains, but I am not in entire agreement that the openshift build experience is 100% applicable.

section will explain, manipulation of environment variable settings on the resulting k8s Container objects exists
for Shipwright.

How to best allow Shipwright users to manage in a first class way from our API environment variables that are ultimately set on k8s 
Containers is what this enhancement proposal will address.

## Motivation

The various image management tools Shipwright encapsulates with 'BuildStrategies' all allow for tuning of their behavior
via environment variables.  The names and supported values of those variables vary between tools, and sometimes 
between different versions of those tools.  So a lot of variability we need to account for.

The same goes for source code management tools, tools for retrieving content via socket connections, both of which are 
used in Shipwright.

So supporting our users via our API and related features to manage environment variables for those tools makes a lot of
sense.

### Goals

- Shipwright API needs to articulate a first class way for specifying those environment variables approach for both 
  administrators and individual developers (i.e. authors/owners of the BuildStrategies vs. authors owners of the Builds 

>   and BuildRuns) to specify those environment variables.

Author: @sbose78 @SaschaSchwarze0 refresh my memory on your respective takes on administrative controls.  IIRC, between the two of you, goals you desired included
- prevention of overriding in build/buildruns of env var k/v defined in the strategy
- specification of finite sets of env vars that could be speficied

Are those incorrect?  Or did I miss something?

As I noted in https://github.com/shipwright-io/build/pull/726#issuecomment-833663332 @adambkaplan *perhaps* a suitable compromise is laying out a long term road map of a complete set of features in this space in this EP, but implementing them in stages.

But some of that is dependent on making sure of the desire amongst the other reviewers on these administrative controls.

>   and BuildRuns) to specify those environment variables.

Reviewer 2: Correct, there are environment variables specified in a build strategy that must not be change in a build or buildrun. And yes, it makes sense that build strategy authors define which environment variables can be set. I mainly focus on the Buildpacks scenario where the different Buildpacks are configurable using environment variables that start with `BP`, such as [BP_MRI_VERSION for Ruby](https://github.com/paketo-buildpacks/mri#mri-configurations).

- Use of the existing features Tekton provides to enrich our users' management of environment variables needs to be 
  evaluated and exploited for code reuse.
- Maintain the currently Shipwright approach of treating Tekton as an implementation detail, and try to not leak Tekton
  API into our API.

### Non-Goals

- Any needed change to integrate Shipwright with Tekton parameters is handled by the [current 'Parameterize Build Strategies' EP](https://github.com/shipwright-io/build/pull/697)
  or follow up work that stems from that EP.  Until then, only static content of any environment variable value related 
  fields is supported.  In other words, the `$(params...)` syntax seen in Tekton today.
- The EP is *NOT* about environment variables that will be set in the final image.  It is about providing environment variables
  to the Steps/Containers and making those environment variables available to the tools called from those Steps/Containers
  as part of building the image.

## Proposal

First, some role / actor terminology and detail (nothing new most likely for Shipwright community members):

> - A `developer` or `end user` is the person who has been provisioned a namespace, or shares a namespace with some teammates, 

Author: part of next push

}
```

This proposal is making the simplifying assumption that ultimately setting the `Env []EnvVar` field in the 
underlying Tekton `Step`/`Container` is sufficient for an initial implementation and scenarios we want to support.
That will help the initial implementation in that the explicit "name" index field helps with default kubernetes 
"path" API operations that tend to occur with devops/gitops flows.  Also, `ValueFrom *EnvVarSource` still allows
us to pull content from `Secrets`, like authentication token.  Many of the tools when building images (including
source code extraction and tools that have to download dependencies) consume authentication tokens via environment
variables.  Storing those token in k8s `Secrets` vs. having to specify them in scripts referenced in `Containers` has 
become a preferred method when building images in k8s native environments.

Next, the question of whether to use the k8s `EnvVar` directly exists.  This proposal suggests creating a new
Shipwright type that inlines the k8s type, in case we want to extend and add new features later that require
additional fields or new types from k8s or Tekton.  So with that, we start with

```go

> type EnvironmentVariable struct {

Reviewer 4: Shorten to `EnvVar` (and `envVar` in our YAML representation?)

Author: I was worried about confusion with the existing k8s structs, but I'll acquiesce, unless others jump in with that similar concern.  I'll give a day or so to see if that occurs before making the switch.

Reviewer 4: We should use k8s namings if they correspond directly to what gets added to a pod spec. Developers won't need to context switch.

might update them with an array of `EnvironmentVariable`

`BuildStrategy` needs some further consideration, in that `BuildStep` already inlines `Container`, which allows specification 
of `EnvVar`.

Those existing `BuildStep` / `Container` environment variables can be the way an administrator could ensure that 
any environment variable exists and has a given setting.  These would be environment variables that could not be 
changed by a `Build` or `BuildRun`.

But in the case where an organization wants to provide a default environment variable setting, but allow their `Build` and
`BuildRun` creators to override the value for that setting (including setting it to the empty string, which in most 
cases would mean "turning off" the environment variable), the `BuildStrategySpec` would have an array of Shipwright
`EnvironmentVariable` added to it.

Lastly, so there are two basic mindsets from an admin perspective:

- Allow any environment variable in general, but restrict a few
- Disallow any environment variable in general, but allow a few

To facilitate both mindsets, we'll add two string arrays to the `BuildStrategySpec`
- `AllowedEnvironmentVariables`

> - `DisallowedEnvironmentVariables`

Reviewer 4: This is perhaps a bit too misleading, since this controls env var overrides at the Build/BuildRun level. Perhaps rename as:

1. `allowedEnvVarOverrides`
2. `blockedEnvVarOverrides`

We should validate that only one of the two are set for a given BuildStrategy/ClusterBuildStrategy. This can be done directly on the generated CRD via OpenAPI - no need for validating admission webhooks.

Author: No those are suppose to be general allow / block lists.  Though initial reaction I do like `blocked...` better than `disallowed...`

Prevention of overrides at the moment is specifying k/v pairs in the existing BuildStep/k8s Container env var array vs. the new array (which in response to another comment you had, minimally needs a rename, if we keep at all).

At this point, it perhaps makes sense to have a common voice to voice / face to face conversation with the current set of reviewers on the "administrative control" themed features.

Based on where we land with that, we can resume this thread as needed.  Where even if we keep these, field names and the surrounding text in the EP perhaps need to be clearer.

Author: > webhook

on the no need for validating webhooks, we may need to cross reference your take here with @SaschaSchwarze0 's https://github.com/shipwright-io/build/pull/726#discussion_r611570295
where part of that I think arose in the recent parameters EP

Each string array list the `EnvVar` names/keys to consider.
 

> If `AllowedEnvironmentVariables` is empty, assume all are allowed, except for those mentioned in `DisallowedEnvironmentVariables`.  

Reviewer 4: ```suggestion
If `AllowedEnvironmentVariables` is empty, assume all environment variable overrides are allowed, except for those mentioned in `DisallowedEnvironmentVariables`.  
```

Author: part of next push

  where values are replaced from any entries previously seeded
- then from `BuildStrategySpec.[]BuildStep`, assuming any names from `BuildStrategySpec.([]string)DisallowedEnvironmentVariable` are not used,
  where values are replaced from any entries previously seeded  

To reiterate a key detail, the above methodology also provide a means of declaring whether an environment variable key/value setting can and connect be overridden. 
If you want to ensure that environment variable key/value settings are *NOT* overridden, specify them in the `BuildStep.`  If you want
to allow them to be overridden, specify them in the `BuildStrategySpec.[]EnvironmentVariable`, `BuildStrategySpec.[]EnvironmentVariable`, `BuildSpec.[]EnvironmentVariable`, or `BuildRunSpec.[]EnvironmentVariable`

Lastly, some yaml to complete the visualization.

`BuildStrategy` with a disallow entry, a spec level default, and a final, always set via a `BuildStep`:

```yaml
apiVersion: shipwright.io/v1alpha1
kind: ClusterBuildStrategy
metadata:
  name: something-that-uses-docker
spec:
  # nobody is allowed to change DOCKER_TLS_VERIFY

>   disallowedEnvironmentVariables:
>     - 'DOCKER_TLS_VERIFY'
>   environmentVariables:
>     # for convenience, we add the version most our users can use, but we'll allow you to override
>     - name: DOCKER_API_VERSION
>       value: '1.19'
>   buildSteps:
>     - command:
>         - /usr/local/bin/docker
>         - build
>       env:
>         - name: DOCKER_CONFIG
>           value: /tekton/home/.docker

Reviewer 4: I am a bit confused by this API. Are `environmentVariables` applied to every build step with the value provided? Can a developer override the `DOCKER_CONFIG` env var?

I would almost rather have build strategy authors define the env vars in line with the build steps, and not worry about (redundantly) declaring default values in `environmentVariables`. Values that can't be overrode should be declared clearly and not rely on hidden behavior. As I understand the behavior in the EP, I think the following would provide more clarity:

```suggestion
  disallowedEnvironmentVariables:
    # Don't allow docker to push to an insecure registry
    - 'DOCKER_TLS_VERIFY'
    # Don't allow the pull/push secrets to be overrode from what Shipwright provides
    - 'DOCKER_CONFIG'
  buildSteps:
    - command:
        - /usr/local/bin/docker
        - build
      env:
        - name: DOCKER_CONFIG
          value: /tekton/home/.docker
            environmentVariables:
       # for convenience, we add the version most our users can use, but developers can override at build time.
       - name: DOCKER_API_VERSION
         value: '1.19'
```

Author: Yes they are applied to build steps, but then those k/v paris could be overwritten by the same keys specified in builds or build runs.

Where as the items specified directly in the BuildStep cannot be overridden.

But assuming we go with this "administrative control", let's dive into the alternative noted by @SaschaSchwarze0 in https://github.com/shipwright-io/build/pull/726#issuecomment-833824197

```
steps:
  - name: my-build-tool
    env:
      - name: MY_FIXED_ENVVAR
	    value: "fixed value"
	    overwritable: false             # default: true
      - name: MY_DEFAULT_VALUE_ENVVAR
	    value: "default value"
```

Well, if you consider the current def of BuildStep:

```
type BuildStep struct {
	corev1.Container `json:",inline"`
}
```

To achieve that yaml, we need to either 

a) create a `shipwright.Container` type, with all the existing `corev1.Container` fields, but with the `envVar` array field change to have to be an array field of a new type that has the `overrwritable` flag.

That seems like a more disruptive change.  And fragile given the field complexity of `corev1.Container`.

or 

b) create a 

```
type BuildStep struct {
   corev1.Container `json:",inline"`
   EnvVarOverwritable   map[string]bool
}

```
Where the key to that map is the env var name, and then true/false indicates whether the build run or build can overwrite.

Of those two, at least for me b) is the more tolerable of the two.  But it is a slight departure from the yaml @SaschaSchwarze0 noted.

When I originally wrote this, b) was not as palatable as what is currently in the EP, but I could be talked into switching to b) now if the current form is too lacking in intuitiveness.

features that exist with the underlying tools used for image management, source code management, or data transfer, are used when
Shipwright runs in the cluster.

### Implementation Details/Notes/Constraints [optional]

### Risks and Mitigations

Coordination with the parameterization enhancement work is an added bit of complexity.

## Design Details

### Test Plan

This requires new unit and integration tests.  In conjunction, documentation needs to be updated with explanations
and examples.

### Graduation Criteria

Should be part of any release before our v1.0.0

### Upgrade / Downgrade Strategy

Nothing around environment variables exists yet in the API.  For changes after the initial implemenetation of 
this proposal, see the references to the kubernetes recommendations elsewhere in this proposal.

### Version Skew Strategy

If this feature does not make v1alpha1, or if we change it in later versions, then the kubernetes recommendations around the use of annotations
to deal with round trip transfer of `BuildStrategy`, `Build`, and `BuildRun` referenced in the [open questions](#open-questions-optional)
section will be needed.

## Implementation History

## Drawbacks

> 

Reviewer 4: What are the drawbacks to this approach? Does our API and planned implementation preclude some future behavior?

Author: part of next push

With that, a new Shipwright type would be:

```go
type EnvironmentVariable struct {
   corev1.EnvVar
   
   // DefaultValue if set is the value applied if neither EnvVar.Value or EnvVar.ValueFrom have a value even 
   // after any Parameter variable substitution.
   // +optional
   DefaultValue  string
}
```

Based on user response, adding `DefaultValue` to our `EnvironmentVariable` wrapper of k8s `EnvVar` is plausible as a
clean, new field only, API extension

### Config for order of precedence

Also, in earlier iterations of this EP, configuration for "developer centric" vs. "cluster admin / ops centric" order 
of precedence was consdered.

A proposed default would be "developer centric":
- apply `BuildStrategy` envs first
- then `Build`
- then `BuildRun`

Then, via the [current global config of controller env vars](https://github.com/shipwright-io/build/blob/v0.4.0/docs/configuration.md),
or if that current approach gets converted to a `ConfigMap`, per (https://github.com/shipwright-io/build/issues/651)[https://github.com/shipwright-io/build/issues/651], or if the community decides to create a `ConfigMap` for each

> logical piece of "feature configuration", we have a setting that allows a switch to "ADMINISTRATOR_CENTRIC" order of precedence,

Reviewer 4: Agree that we shouldn't do this - stick to one behavior only.

changed by a `Build` or `BuildRun`.

But in the case where an organization wants to provide a default environment variable setting, but allow their `Build` and
`BuildRun` creators to override the value for that setting (including setting it to the empty string, which in most 
cases would mean "turning off" the environment variable), the `BuildStrategySpec` would have an array of Shipwright
`EnvironmentVariable` added to it.

Lastly, so there are two basic mindsets from an admin perspective:

- Allow any environment variable in general, but restrict a few
- Disallow any environment variable in general, but allow a few

To facilitate both mindsets, we'll add two string arrays to the `BuildStrategySpec`
- `AllowedEnvironmentVariables`
- `DisallowedEnvironmentVariables`

Each string array list the `EnvVar` names/keys to consider.
 
If `AllowedEnvironmentVariables` is empty, assume all are allowed, except for those mentioned in `DisallowedEnvironmentVariables`.  
Otherwise, the list is the list, and any environment variables in `BuildStep`, `Build`, and `BuildRun` must be vetted against this list.  

> Initially, no wildcards or regexp, but perhaps a feature to consider in the future.

Reviewer 2: I think you can remove this line as you later mention

> Each entry of the allow and disallow arrays need to minimally support wildcards, if not regular expressions.

Author: agreed
